server/src/main/scala/org/http4s/server/middleware/VirtualHost.scala

```diff
@@ -0,0 +1,41 @@
+package org.http4s
+package server
+package middleware
+
+import headers.Host
+import Status.BadRequest
+
+/** Middleware for virtual host mapping
+  *
+  * The `VirtualHost` middleware allows multiple services to be mapped
+  * based on the [[Host]] header of the [[Request]]. A default service
+  * is specified which will handle cases where the [[Host]] header is
+  * not present (should only happen for HTTP/1.0 requests) or if none
+  * of the hosts match.
+  * If the port is not specified by the client _or_ the server, it is
+  * ignored because the middleware cannot know the default port for the
+  * underlying transport.
+  */
+object VirtualHost {
+
+  /** Specification of the virtual host */
+  case class HostService(service: HttpService, host: String, port: Option[Int] = None)
+
+  def apply(default: HostService, rest: HostService*): HttpService = {
+
+    def check(host: Host, hs: String, port: Option[Int]): Boolean =
+      host.host == hs && (port.isEmpty || host.port.isEmpty || host.port.get == port.get)
```

Comment:
By adding more fields or attributes to the `Request`? That might not be a bad idea. Currently the backends add the remote address to the attributes map, but I can see this going either way.